src/search/landmarks/landmark_count_heuristic.cc

```diff
@@ -237,46 +242,49 @@ bool LandmarkCountHeuristic::generate_helpful_actions(
                 continue;
             FactProxy fact_proxy = effect.get_fact();
             LandmarkNode *lm_node = lgraph->get_node(fact_proxy.get_pair());
-            if (lm_node && landmark_is_interesting(state, reached, *lm_node)) {
+            if (lm_node && landmark_is_interesting(
+                    state, reached, *lm_node, all_landmarks_reached)) {
                 if (lm_node->get_landmark().disjunctive) {
-                    ha_disj.push_back(op_id);
+                    preferred_operators_disjunctive.push_back(op_id);
                 } else {
-                    ha_simple.push_back(op_id);
+                    preferred_operators_simple.push_back(op_id);
                 }
             }
         }
     }
-    if (ha_disj.empty() && ha_simple.empty())
-        return false;
 
     OperatorsProxy operators = task_proxy.get_operators();
-    if (ha_simple.empty()) {
-        for (OperatorID op_id : ha_disj) {
+    if (preferred_operators_simple.empty()) {
+        for (OperatorID op_id : preferred_operators_disjunctive) {
             set_preferred(operators[op_id]);
         }
     } else {
-        for (OperatorID op_id : ha_simple) {
+        for (OperatorID op_id : preferred_operators_simple) {
             set_preferred(operators[op_id]);
         }
     }
-    return true;
 }
 
 bool LandmarkCountHeuristic::landmark_is_interesting(
-    const State &state, const LandmarkNodeSet &reached, LandmarkNode &lm_node) const {
-    /* A landmark is interesting if it hasn't been reached before and
-     its parents have all been reached, or if all landmarks have been
-     reached before, the LM is a goal, and it's not true at moment */
-
-    int num_reached = reached.size();
-    if (num_reached != lgraph->get_num_landmarks()) {
-        if (reached.find(&lm_node) != reached.end())
-            return false;
-        else
-            return !check_node_orders_disobeyed(lm_node, reached);
+    const State &state, const BitsetView &reached,
+    LandmarkNode &lm_node, bool all_lms_reached) const {
+    /*
+      A landmark is interesting if it hasn't been reached before and
+      its parents have all been reached, or if all landmarks have been
+      reached before, the LM is a goal, and it's not true at moment.
+    */
+
+    if (all_lms_reached) {
+        const Landmark &landmark = lm_node.get_landmark();
+        return landmark.is_true_in_goal && !landmark.is_true_in_state(state);
+    } else if (reached.test(lm_node.get_id())) {
+        return false;
+    } else {
+        return all_of(lm_node.parents.begin(), lm_node.parents.end(),
+                      [&](const pair<LandmarkNode *, EdgeType> parent) {
+                          return reached.test(parent.first->get_id());
+                      });
```

Comment:
To me, based on the comment above, this would be clearer as a two-case distinction rather than a three-case distinction, with the second case being "!reached.test(...) && all_of(...)". Common advice would perhaps also be to have the if/else order match the order of the two cases in the explaining comment; currently it's opposite. Since another common advice is to avoid negative conditions, this perhaps would entail changing the order in the comment, not in the code.